### PR TITLE
BUG: GeoDataFrame CRS loss during __setitem__ 

### LINKED ISSUE #1372: BUG: GeoDataFrame CRS loss during __setitem__ on pandas master
Due to some recent changes on pandas master, GeometryArray loses its CRS during `__setitem__`. In the following example from our tests, resulting CRS should be the one of passed array (4326).

```py
from geopandas.array import from_shapely

geoms = [Point(0, 0), Point(1, 1)]
df = GeoDataFrame({"col1": [0, 1]}, geometry=geoms, crs=27700)
df["geometry"] = from_shapely(geoms, crs=4326)

df.crs
<Projected CRS: EPSG:27700>
...
```

Related issues on pandas side pandas-dev/pandas#33457, pandas-dev/pandas#32960

### LINKED ISSUE #1960: BUG: GeoDataFrame.crs not updated if geometry GeoSeries is reprojected manually
Consider following example.

```py
gdf = geopandas.read_file(geopandas.datasets.get_path("nybb"))
gdf.crs

<Projected CRS: EPSG:2263>
...

gdf["geometry"] = gdf.geometry.to_crs(4326)
gdf.crs

<Projected CRS: EPSG:2263>
...
```

GeoDataFrame doesn't load the CRS of its active geometry column if it is re-projected in this way. Assigning re-projected GeoSeries as an attribute (`gdf.geometry = gdf.geometry.to_crs(4326)`) works as expected.

`gdf.crs` should directly return the CRS of the active geometry column in this case.

### PR DESCRIPTION
**Pull request by m-richards**
This fixes #1372, and by association fixes #1960

### CODE REVIEW

File: geopandas/tests/test_crs.py [hunk 1]
```diff
@@ -293,6 +295,30 @@ def test_dataframe(self):
         assert df.geometry.crs == self.wgs
         assert df.geometry.values.crs == self.wgs
 
+        # new geometry with set CRS has priority over GDF CRS -
```
Comment (jorisvandenbossche) [inline geopandas/tests/test_crs.py:298]: This test is already quite long, so can you maybe add it as a new test? (eg `test_dataframe_setitem`)

**Review by jorisvandenbossche — COMMENTED** [review]
@m-richards Thanks a lot for looking into this!

I think this implementation might give a problem when the geometry column you are setting is not the default "active geometry column". Because in that case, we should not set the GeoDataFrame's crs to the crs of the new values.

Can you add a case for that to the new test to verify this?

Something like:

```
arr = from_shapely(self.geoms)
df = GeoDataFrame({"col1": [1, 2], "geometry": arr, "other_geom": arr}, crs=4326)
df["other_geom"] = arr.to_crs(27700)
assert df.crs == 4326
assert df.geometry.crs == 4326
assert df["geometry"].crs == 4326
assert df["other_geom"].crs == 27700
```

File: geopandas/tests/test_crs.py (hunk 1 above)
Comment (m-richards) [reply geopandas/tests/test_crs.py:298]: Thanks for the feedback, they're both good ideas, I'll get onto that.

**Comment by m-richards** [comment]
> I think this implementation might give a problem when the geometry column you are setting is not the default "active geometry column". Because in that case, we should not set the GeoDataFrame's crs to the crs of the new values.

I've added a test for this, which already passes. The assignment `self._crs = value.crs` sits inside a conditional for the active geometry column so I'm pretty sure that is why there are no issues with this (below, or line 1320 of geodataframe.py). 

```python
  if not pd.api.types.is_list_like(key) and key == self._geometry_column_name:
            if pd.api.types.is_scalar(value) or isinstance(value, BaseGeometry):
                value = [value] * self.shape[0]
            try:
                value = _ensure_geometry(value, crs=self.crs)
                self._crs = value.crs
```

**Comment by jorisvandenbossche** [comment]
>  sits inside a conditional for the active geometry column so I'm pretty sure that is why there are no issues with this

Ah, yes indeed, I missed the `key == self._geometry_column_name` in the conditional.

**Review by martinfleis — APPROVED** [review]
Looks good to me. Thanks!

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request merged**